File: Problems/No.1620_Coordinate_With_Maximum_Network_Quality/coordinate_with_maximum_network_quality.cpp
class Solution {
public:
    vector<int> bestCoordinate(vector<vector<int>>& towers, int radius) {
        
        // parameters
        int MAX_x = 0, min_x = INT_MAX;
        int MAX_y = 0, min_y = INT_MAX;
        int MAX_quality = 0;
        std::vector<int> highest_quality = {0, 0};

        // start computing
        // calculate the edge points of the bounding box
        for (auto &coordinate: towers) {

            int tower_x = coordinate[0], tower_y = coordinate[1];
            MAX_x = std::max(MAX_x, tower_x);
            MAX_y = std::max(MAX_y, tower_y);
            min_x = std::min(min_x, tower_x);
            min_y = std::min(min_y, tower_y);

        }

        // find the highest quality from the inside of the bounding box
        for (int x=min_x; x<=MAX_x; x++) 
            for (int y=min_y; y<=MAX_y; y++) {

                int curr_quality = 0;
                for (auto &tower: towers) {

                    float distance = std::sqrt(std::pow(x-tower[0], 2) + std::pow(y-tower[1], 2));
                    if (distance > radius) continue;
                    curr_quality += (tower[2]/(1 + distance));

                }
                if (curr_quality > MAX_quality) {

                    MAX_quality = curr_quality;
                    highest_quality = {x, y};

                }

            }

        return highest_quality;

    }
};

File: Problems/No.1620_Coordinate_With_Maximum_Network_Quality/coordinate_with_maximum_network_quality.py
class Solution(object):
    def bestCoordinate(self, towers, radius):
        """
        :type towers: List[List[int]]
        :type radius: int
        :rtype: List[int]
        """

        # parameters
        MAX_x, min_x = 0, float('inf')
        MAX_y, min_y = 0, float('inf')
        MAX_quality = 0
        highest_quality = [0, 0]

        # start computing
        # calculate the edgepoints of the bounding box
        for [tower_x, tower_y, _] in towers:
            MAX_x = max(MAX_x, tower_x)
            MAX_y = max(MAX_y, tower_y)
            min_x = min(min_x, tower_x)
            min_y = min(min_y, tower_y)
        
        # find the highest quality from the inside of the bounding box
        for x in range(min_x, MAX_x+1):
            for y in range(min_y, MAX_y+1):
                curr_quality = 0
                for [tower_x, tower_y, tower_quality] in towers:
                    distance = math.sqrt((x-tower_x)**2 + (y-tower_y)**2)
                    if (distance > radius):
                        continue
                    curr_quality += (tower_quality//(1 + distance))
                if (curr_quality > MAX_quality):
                    MAX_quality = curr_quality
                    highest_quality = [x, y]
        
        return highest_quality
